File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        vec_r = [0, 0, -1, 1]
        vec_c = [-1, 1, 0, 0]
        visited = [ [0] * (len(grid[0])) for _ in range(len(grid)) ]

        def dfs(r: int, c: int):
            if visited[r][c] == 1 or grid[r][c] == "0":
                return;
            visited[r][c] = 1
            for i in range(4):
                new_r = r + vec_r[i]
                new_c = c + vec_c[i]
                if new_r >= 0 and new_r < len(grid) and new_c >=0 and new_c < len(grid[0]):
                    if visited[new_r][new_c] == 0 and grid[new_r][new_c] == "1":
                        dfs(new_r, new_c) 
        
        count = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if visited[i][j] == 1 or grid[i][j] == "0":
                    continue
                dfs(i, j)
                count += 1

        return count

File: 200-number-of-islands/number-of-islands.java
class Solution {
    int[][] visited;
    int[] vec_r = new int[]{0, 0, -1, 1};
    int[] vec_c = new int[]{-1, 1, 0, 0};
    int count = 0;

    public int numIslands(char[][] grid) {
        visited = new int[grid.length][grid[0].length];

        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[0].length; c++) {
                if (visited[r][c] == 1 || grid[r][c] == '0') continue;
                dfs(r, c, grid);
                count += 1;
            } 
        }

        return count;
    }

    private void dfs(int r, int c, char[][] grid) {
        if (visited[r][c] == 1 || grid[r][c] == '0') return;
        visited[r][c] = 1;
        for (int i = 0; i < 4; i++) {
            int new_r = r + vec_r[i];
            int new_c = c + vec_c[i];
            if (new_r < 0 || new_r >= grid.length || new_c < 0 || new_c >= grid[0].length) continue;
            if (visited[new_r][new_c] == 1 || grid[new_r][new_c] == '0') continue;
            dfs(new_r, new_c, grid);
        }
    }
}